tensorflow/contrib/seq2seq/python/kernel_tests/seq2seq_test.py

```diff
@@ -0,0 +1,114 @@
+                                         input_size])
+        decoder_inputs = tf.constant(0.4, shape=[decoder_sequence_length,
+                                                 batch_size,
+                                                 decoder_input_size])
+        decoder_length = tf.constant(decoder_sequence_length, dtype=tf.int32,
+                                     shape=[batch_size,])
+
+        # Define model
+        encoder_outputs, encoder_state = tf.nn.dynamic_rnn(
+            cell=tf.nn.rnn_cell.GRUCell(encoder_size), inputs=inputs,
+            dtype=tf.float32, time_major=True, scope="rnn")
+        decoder_cell = tf.nn.rnn_cell.GRUCell(decoder_size)
+        decoder_cell = tf.nn.rnn_cell.OutputProjectionWrapper(decoder_cell,
+                                                              output_size)
+
+        # Train decoder
+        decoder_fn_train = tf.contrib.seq2seq.simple_decoder_fn_train(
+            encoder_state=encoder_state)
+        decoder_outputs_train, decoder_state_train = (
+          tf.contrib.seq2seq.dynamic_rnn_decoder(
+              cell=decoder_cell,
+              decoder_fn=decoder_fn_train,
+              inputs=decoder_inputs,
+              sequence_lengths=decoder_length,
+              time_major=True))
+
+        # Inference decoder
+        varscope.reuse_variables()
+        decoder_fn_inference = tf.contrib.seq2seq.simple_decoder_fn_inference(
+            encoder_state=encoder_state,
+            embeddings=decoder_embeddings,
+            start_of_sequence_id=start_of_sequence_id,
+            end_of_sequence_id=end_of_sequence_id,
+            maximum_length=decoder_sequence_length,
+            dtype=tf.int32)
+        decoder_outputs_inference, decoder_state_inference = (
+          tf.contrib.seq2seq.dynamic_rnn_decoder(
+              cell=decoder_cell,
+              decoder_fn=decoder_fn_inference,
+              time_major=True))
+
+        # Run model
+        tf.initialize_all_variables().run()
+        decoder_outputs_train_res, decoder_state_train_res = sess.run(
+            [decoder_outputs_train, decoder_state_train])
+        decoder_outputs_inference_res, decoder_state_inference_res = sess.run(
+            [decoder_outputs_inference, decoder_state_inference])
+
+        # Assert outputs
+        self.assertEqual((decoder_sequence_length, batch_size, output_size),
+                         decoder_outputs_train_res.shape)
+        self.assertEqual((batch_size, decoder_size),
+                         decoder_outputs_inference_res.shape[0:2])
+        self.assertEqual((batch_size, decoder_size),
+                         decoder_state_train_res.shape)
+        self.assertEqual((batch_size, decoder_size),
+                         decoder_state_inference_res.shape)
+        # The dynamic decoder might end earlier than `maximal_length`
+        # under inference
+        true_value = (decoder_sequence_length>=
```

Comment:
Okay, I have been working some on this, here are my thoughts.

I assume you want something in the regards of `decoder_outputs_inference == decoder_outputs_train`, given they have access to the same embedding.

This is really tricky because of the following reasons:

**The decoder picks random embeddings at every iteration**

We can solve this by taking the decoder_inference's output, argmax it and pick the same embeddings for the to match `decoder_inputs` (training decoder).

**The decoder randomly stops whenever it finds a EOS symbol, and it doesn't tell you when it stopped**

The problem here is that whenever it finishes (`True` in the `done` vector), `raw_rnn` just starts copying states.
Given that our current solution does not return the `done` vector to tell when it stopped, we can't assess when it started copying states. We could do some hack analyzing the `decoder_output_inference` looking for when it started copying, but I don't think this is the optimal solution.

Perhaps we should consider returning the `done` vector from the `dynamic_rnn_decoder`, but not just a boolean vector, a vector with information on when every sequence in the batch stopped. So like a reverse engineered `sequence_lengths`.
Afterwards we would use this as input for the training decoder, which should allow this test.

Or, simply leave this test out ..?